### DISCUSSION TITLE
How to encode register addresses - Modbus over TCP

### DISCUSSION THREAD

**Discussion by bergheil** [aldas/modbus-tcp-client#118] [Q&A]
Hi, I'm using the library to read from modbus over tcp.
I'm able to connect but I cannot receive data after request. The registers addresses are indicated as the following:

```
3x0220
3x0221
3x0222
3x0230
4x0240

```

My code is taken from example_cli_poller.php example.
```

indirizzo = 'tcp://192.168.178.178:8000';
$requests = ReadRegistersBuilder::newReadHoldingRegisters($indirizzo)    
    ->int16(220, 'STATO MACCHINA: 0=CICLO OFF  1=CICLO ON')  
    ->build(); // returns array of 3 requests

// Install: 'composer require react/socket:^1.6'
// NB: install PHP extension ('ev', 'event' or 'uv') if the concurrent socket connections are more than 1024.

$loop = React\EventLoop\Factory::create();

$n = 100;
$loop->addPeriodicTimer(1.0, function () use ($loop, &$n, $requests) {
    echo microtime(true) . ": TICK! remaining {$n} seconds" . PHP_EOL;

    foreach ($requests as $request) {
        $connector = new React\Socket\Connector($loop, array(
            'dns' => false,
            'timeout' => 2.2
        ));

        $connector->connect($request->getUri())->then(
            function (React\Socket\ConnectionInterface $connection) use ($request) {
                $receivedData = b'';

                echo microtime(true) . ": connected to {$request->getUri()}" . PHP_EOL;
                $connection->write($request);

                // wait for response event
                $connection->on('data', function ($data) use ($connection, $request, &$receivedData) {
                    // there are rare cases when MODBUS packet is received by multiple fragmented TCP packets and it could
                    // take PHP multiple reads from stream to get full packet. So we concatenate data and check if all that
                    // we have received makes a complete modbus packet.
                    // NB: `Packet::isCompleteLength` is suitable only for modbus TCP packets
                    $receivedData .= $data;
                    if (Packet::isCompleteLength($receivedData)) {
                        echo microtime(true) . ": uri: {$request->getUri()}, complete response: " . print_r($request->parse($receivedData), true) . PHP_EOL;
                        $connection->end();
                    } else {
                        echo microtime(true) . ": uri: {$request->getUri()}, partial response: " . print_r($data, true) . PHP_EOL;
                    }
                });
                $connection->on('error', function ($data) use ($connection, $request) {
                    echo microtime(true) . ": uri: {$request->getUri()}, Error during connection! error: " . print_r($data, true) . PHP_EOL;
                    $connection->end();
                });
            },
            function (Exception $error) use ($request) {
                echo microtime(true) . ": uri: {$request->getUri()}, failed to connect! error: " . $error->getMessage() . PHP_EOL;
            });
    }
    if ($n <= 0) {
        $loop->stop();
    }
    $n--;
});
$loop->run();
```

The output is 

```
1663226895.4214: TICK! remaining 89 seconds
1663226895.4747: connected to tcp://192.168.178.178:8000
1663226896.4225: TICK! remaining 88 seconds
1663226896.4768: connected to tcp://192.168.178.178:8000
```

Can anyone help me to understand how to read from the machine?

Bye

**Comment by aldas** [comment]
Which `unit id`/`slave id` you should use? By default `0` is used and you can define as second argument at [`ReadRegistersBuilder::newReadHoldingRegisters` ](https://github.com/aldas/modbus-tcp-client/blob/ac8661a693aed70d542a3c4d10cce08ea1ab88b5/src/Composer/Read/ReadRegistersBuilder.php#L46)

If this does not work please try to use https://github.com/aldas/modbus-tcp-client/wiki/Debugging-guide to pinpoint your problem. Mostly that second block with trying another client to see if it makes a difference.

**Reply by francescopfc** [reply]
Thanks very much. Changing the unit id now I'am able to connect. 
The next step is understand how to read the machine information. I have the attached registry list.

I tried with this code but I do not read any information:

```
$requests = ReadRegistersBuilder::newReadHoldingRegisters($indirizzo, $unitId)    
    ->int16(220, 'STATO MACCHINA: 0=CICLO OFF  1=CICLO ON')
    ->build(); 

```

Thanks
![image](https://user-images.githubusercontent.com/83782370/198079894-125cf8ef-71aa-4d1b-b1a6-cf6e1ca8cf34.png)

**Reply by bergheil** [reply]
Sorry, I used a different github account.

**Reply by aldas** [reply]
maybe `0220` is in hex? `0x0220`

**Reply by bergheil** [reply]
Probably but I tried in decimal.

Using the client modpoll and the php poller I got the same error.

`./modpoll -m tcp -a 1 -r 544 -c 1 -p 8000 -t 4 192.168.178.178`


Copyright (c) 2002-2021 proconX Pty Ltd
Visit https://www.modbusdriver.com for Modbus libraries and tools.

```
Protocol configuration: MODBUS/TCP, FC3
Slave configuration...: address = 1, start reference = 544, count = 1
Communication.........: 192.168.178.178, port 8000, t/o 1.00 s, poll rate 1000 ms
Data type.............: 16-bit register, output (holding) register table

-- Polling slave... (Ctrl-C to stop)
Illegal Data Address exception response!
-- Polling slave... (Ctrl-C to stop)
Illegal Data Address exception response!
-- Polling slave... (Ctrl-C to stop)
Illegal Data Address exception response!
-- Polling slave... (Ctrl-C to stop)

```

**Reply by aldas** [reply]
well, if `modpoll` also fails I do not know.  Maybe unit id is something other than 1, maybe documentation is wrong and maybe FC4 would work for these addresses. Try to find that device manual and see if it has some details/remarks about addressing.

NB: try  adding -1 and +1 to addresses. there is a difference in modpud PDU address and "documentation" address. PDU address starts from `0` but the other way of giving address is starting from `1`

**Reply by bergheil** [reply]
Hi, I tried again with modpoll and seems ok. 

`./modpoll -m tcp -a 1 -r 220 -c 1 -p 8000 -t 3 192.168.178.178`

the output is:

modpoll 3.10 - FieldTalk(tm) Modbus(R) Master Simulator
Copyright (c) 2002-2021 proconX Pty Ltd
Visit https://www.modbusdriver.com for Modbus libraries and tools.
```

Protocol configuration: MODBUS/TCP, FC4
Slave configuration...: address = 1, start reference = 220, count = 1
Communication.........: 192.168.178.178, port 8000, t/o 1.00 s, poll rate 1000 ms
Data type.............: 16-bit register, input register table

-- Polling slave... (Ctrl-C to stop)
[220]: 0
-- Polling slave... (Ctrl-C to stop)
[220]: 0
-- Polling slave... (Ctrl-C to stop)

```


But in the above example example_cli_poller.php I canno read the value of the register. Can you help me?

Thanks

**Reply by aldas** [reply]
seems that FC4 is working for you. `example_cli_poller.php` uses FC3. So change this line

https://github.com/aldas/modbus-tcp-client/blob/73695f6b6b71dfc43ed9c76b3430706d825901d4/examples/example_cli_poller.php#L13

to
```
$requests = ReadRegistersBuilder::newReadInputRegisters('tcp://127.0.0.1:5022')
```
